design/worker-config-drive.md

```diff
@@ -0,0 +1,196 @@
+host plays (master, worker, etc.).
+
+In order to bring a host into the cluster as a worker we have to give
+it the correct Ignition configuration as it boots.  The target images
+we are using for nodes in an OpenShift (and eventually Kubernetes)
+cluster are RHEL CoreOS images with Ignition configured to look for
+its data in `/openstack/latest` in the config drive. This setting is
+built in to the image, and is not currently configurable.
+
+The Ignition file contents are stored in a Secret within the
+kubernetes database because they contain certificate authority
+information. There are different Ignition settings for worker and
+master nodes, so something will need to know the intended role of each
+host being provisioned in order to select the correct settings.
+
+Something needs to build a config drive image containing the ignition
+file and the baremetal operator needs to pass that formatted image to
+Ironic as part of provisioning the host.
+
+In order to avoid having the baremetal operator coupled tightly to any
+specific images, something that knows about the expectations of the
+target image being deployed (especially the path where the Ignition
+file should be written within the config drive image) should build the
+config drive image, and pass it to the baremetal operator via the host
+along with the other provisioning instructions like the image name and
+checksum.
+
+The component best situated to orchestrate all of those decisions
+currently is the baremetal actuator in the cluster-api provider.  The
+actuator knows which images to provision, since this is built into the
+actuator's settings. It can also know the expected layout of the
+config drive for each image.  And the actuator knows the role of the
+host being allocated to a Machine, so it can select the right Ignition
+configuration to include in the config drive image.
+
+In order to protect the content of the config drive image, the
+actuator should store it as a new Secret, and pass only a reference to
+the baremetal operator when provisioning a host.
+
+## Motivation
+
+### Goals
+
+- Avoid having the baremetal operator tightly coupled to any specific
+  image.
+- Avoid having the baremetal operator tightly coupled to provisioning
+  hosts to become nodes in the cluster.
+- Avoid leaking secrets when passing the config drive to the baremetal
+  operator.
+
+### Non-Goals
+
+N/A
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+Under OpenShift, the worker settings come from the contents of a
+secret (`openshift-machine-api/worker-user-data`), which holds a
+base64-encoded copy of the Ignition JSON data.
+
+Before allocating a host to be a worker, the actuator will identify
+the Secret containing the worker settings, access its contents, and
+create an ISO image with the data in the right layout. It will then
+read the ISO image into memory and compress it, before writing it to a
+new Secret. It will include a reference to that second Secret with the
```

Comment:
This makes it sound like the actuator is going to do create the ISO and compress it - why wouldn't you use gophercloud/utils, which will create the exact format needed by the Ironic API? Or, depending on the Ironic version, just send it the ignition JSON and [have it create the image](https://storyboard.openstack.org/#!/story/2005083)?

Reply:
Here I'm talking about the responsibility for something doing it. The implementation details of how that works is left up to the person writing that code. I think using the utils module you write makes a ton of sense.